oracle/src/currency.rs

```diff
@@ -0,0 +1,251 @@
+#![allow(clippy::upper_case_acronyms)]
+
+use crate::Error;
+use runtime::{CurrencyId, FixedPointNumber, FixedPointTraits::*, FixedU128, TryFromSymbol};
+use serde::{de::Error as _, Deserialize, Deserializer};
+use std::{
+    convert::TryInto,
+    fmt::{self, Debug},
+    str::FromStr,
+};
+
+pub trait ExchangeRate {
+    fn invert(self) -> Self;
+}
+
+impl ExchangeRate for f64 {
+    fn invert(self) -> Self {
+        1.0 / self
+    }
+}
+
+pub trait CurrencyInfo {
+    fn name(&self) -> &str;
+    fn symbol(&self) -> &str;
+    fn decimals(&self) -> u8;
+}
+
+macro_rules! create_currency {
+    ($(#[$meta:meta])*
+	$vis:vis enum Currency {
+        $($(#[$vmeta:meta])* $symbol:ident($name:expr, $deci:literal),)*
+    }) => {
+		$(#[$meta])*
+		$vis enum Currency {
+			$($(#[$vmeta])* $symbol,)*
+		}
+
+        $(
+            #[allow(dead_code)]
+            pub const $symbol: Currency = Currency::$symbol;
+        )*
+
+		impl CurrencyInfo for Currency {
+			fn name(&self) -> &str {
+				match self {
+					$(Currency::$symbol => $name,)*
+				}
+			}
+			fn symbol(&self) -> &str {
+				match self {
+					$(Currency::$symbol => stringify!($symbol),)*
+				}
+			}
+			fn decimals(&self) -> u8 {
+				match self {
+					$(Currency::$symbol => $deci,)*
+				}
+            }
+		}
+
+        impl FromStr for Currency {
+            type Err = Error;
+            fn from_str(symbol: &str) -> Result<Self, Self::Err> {
+                let uppercase_symbol = symbol.to_uppercase();
+                // try hardcoded currencies first
+                match uppercase_symbol.as_str() {
+                    $(stringify!($symbol) => Ok(Currency::$symbol),)*
+                    _ => Err(Error::InvalidCurrency),
+                }
+            }
+        }
+    }
+}
+
+create_currency! {
+    #[derive(PartialEq, Eq, Copy, Clone, Debug)]
+    #[repr(u8)]
+    pub enum Currency {
```

Comment:
I don't like this duplication from the parachain.. Nor the hardcoding of these `USDT`, `USD`, `AUSD` currencies. Imo we should be able to add new currencies without having to recompile the oracle

Reply:
I was also not taken by this (see my comment above) but I have since moved the currencies into the configuration file to make it easy to extend the definitions whenever we add new currencies.